dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionAnalyticsDataFetcherTest.java

```diff
@@ -0,0 +1,489 @@
+            .withPeriods( periods )
+            .withDataDimensionItems( attributeOptionItems )
+            .withOrganisationUnits( partitions.get( 2 ) )
+            .build();
+
+        gridWithAttributeOptionsA = newGridWithAttributeOptions();
+
+        gridWithAttributeOptionsA.addRow()
+            .addValue( periodA.getIsoDate() )
+            .addValue( programIndicatorA.getUid() )
+            .addValue( orgUnits.get( TEST_A ).getUid() )
+            .addValue( attributeOptionComboA.getUid() )
+            .addValue( (Double) 10.0 );
+
+        gridWithAttributeOptionsA.addRow()
+            .addValue( periodA.getIsoDate() )
+            .addValue( programIndicatorB.getUid() )
+            .addValue( orgUnits.get( TEST_A ).getUid() )
+            .addValue( attributeOptionComboA.getUid() )
+            .addValue( (Double) 20.0 );
+
+        gridWithAttributeOptionsA.addRow()
+            .addValue( periodB.getIsoDate() )
+            .addValue( programIndicatorB.getUid() )
+            .addValue( orgUnits.get( TEST_B ).getUid() )
+            .addValue( attributeOptionComboA.getUid() )
+            .addValue( (Double) 30.0 );
+
+        gridWithAttributeOptionsB = newGridWithAttributeOptions();
+
+        gridWithAttributeOptionsB.addRow()
+            .addValue( periodA.getIsoDate() )
+            .addValue( trackedEntityAttributeA.getUid() )
+            .addValue( orgUnits.get( TEST_C ).getUid() )
+            .addValue( attributeOptionComboA.getUid() )
+            .addValue( (Double) 40.0 );
+
+        gridWithAttributeOptionsB.addRow()
+            .addValue( periodA.getIsoDate() )
+            .addValue( trackedEntityAttributeA.getUid() )
+            .addValue( orgUnits.get( TEST_D ).getUid() )
+            .addValue( attributeOptionComboA.getUid() )
+            .addValue( (Double) 50.0 );
+
+        gridWithAttributeOptionsC = newGridWithAttributeOptions();
+
+        gridWithAttributeOptionsC.addRow()
+            .addValue( periodA.getIsoDate() )
+            .addValue( trackedEntityAttributeA.getUid() )
+            .addValue( orgUnits.get( TEST_E ).getUid() )
+            .addValue( attributeOptionComboA.getUid() )
+            .addValue( (Double) 60.0 );
+
+        gridWithoutAttributeOptionsA = newGridWithoutAttributeOptions();
+
+        gridWithoutAttributeOptionsA.addRow()
+            .addValue( periodA.getIsoDate() )
+            .addValue( programAttributeAUid )
+            .addValue( orgUnits.get( TEST_B ).getUid() )
+            .addValue( (Double) 70.0 );
+
+        gridWithoutAttributeOptionsB = newGridWithoutAttributeOptions();
+
+        gridWithoutAttributeOptionsC = newGridWithoutAttributeOptions();
+
+        gridWithoutAttributeOptionsC.addRow()
+            .addValue( periodA.getIsoDate() )
+            .addValue( programAttributeAUid )
+            .addValue( orgUnits.get( TEST_E ).getUid() )
+            .addValue( (Double) 80.0 );
```

Comment:
There is quite some duplication here making this unnecessarily long and hard to see if or what the varying value(s) are for the reoccurring calls.

Reply:
I'm of two minds about this. I agree that helper methods are useful when it makes the intent of the code clearer and I don't mind using them when I think they do that. But when a test is failing, someone new to the test code needs to know exactly what the test conditions are. Sometimes code that is a bit less optimized may be easier to see exactly what is being instrumented. I find in general that a lot of DHIS2 unit tests have a lot of copy-and-paste code that would be totally unacceptable in operational code.

I don't think that helper routines should never be used to optimize copy-and-paste style test code, nor that they should always be used. It's a value judgement each time, where the end goal is always how quickly can someone not familiar with the test code know exactly what is going on.

In this particular case, we're building a lot of _DataQueryParams_ objects, which have their own builder, and _Grid_ objects to which we add rows. I considered your suggestion, and I tried to imagine how I would design them. It's not the case that most of the parameter never change. They change at different times. In this particular case, I felt that helper methods might make it less clear to the reader exactly how these objects are being constructed. We don't want the reader spending a lot of time wondering if an error is in the operational code or in a test helper.

I would also be very interested in any opinions from other group members, or in any further thoughts from you.

Reply:
I understand what you say and I am not here to argue with you or ask for changes that you don't see as an improvement.

I know there is similar duplication in many tests but I in my experience this is not a good thing. I have cleaned up several of these and found that when there is repetitive code there are hidden issues because it gets hard to see the forest for the trees. 

Just now I work on straightening out transaction management and what I especially find with tests repeating basic steps instead of using helpers is that tests do things in the wrong order (use not yet saved entities). This then just works because of hibernate magic queuing the missed operations in the background and squeezing them in where you would not expect them. Except that you might be in a read-only transaction by then. It appears in some places the solution to this has not been to straighten out the test setup but to make read-only DB access use a read-write transaction. This then cascades up and soon lots of things have to be read-write although they clearly only want to read from the database. 

I also can't say that lengthy code helps me understanding what is going on. From a certain length beyond 20-25 lines   instead of being able to read the declared intent of what the test scenario is trying to do there are all these details making me wonder what the tests tries to do, what the important and unimportant values are and so on. It basically means the only way to understand the test is to reverse engineer it and assume intent instead of reading intent and then in case something is odd see if the actual code fails to fulfil what was intended to happen. 
I miss the WHY and the WHAT. There is lots of HOW for some unknown WHAT and WHY. This is what I think helper methods can provide when named accordingly.

Reply:
Thanks. Points well-taken about our tests. I've had the same problem. Perhaps we can do better in the future.

I also think that our code in general is under-commented. I know this is an art where good programmers can have a difference of opinion; on the one hand not wanting comments to get in the way of reading code -- but on the other hand wanting to explain the non-obvious things that some code blocks are doing. I tend to comment more rather than less, but I have cut back in this project because not everyone feels the same. I've tended to put JavaDoc on all private methods I've written as a general rule, because that's less obtrusive to some than interspersing them with the code.

You make a good point about choosing names in the code to make the intent clearer. That can be a good middle ground. I can try to do that more. And I think I may also add more comments especially in my tests in the future to make the WHAT and WHY clearer.

At the very least I'd like to see JavaDoc on all our public classes and methods, rather than having to read all the code to figure out their intent.